src/garage/sampler/ray_batched_sampler.py

```diff
@@ -0,0 +1,216 @@
+"""This is an implementation of an on policy batch sampler.
+
+Uses a data parallel design.
+Included is a sampler that deploys sampler workers.
+
+The sampler workers must implement some type of set agent parameters
+function, and a rollout function
+"""
+from collections import defaultdict
+import pickle
+
+import numpy as np
+import psutil
+import ray
+
+from garage.experiment import deterministic
+from garage.misc import tensor_utils
+from garage.misc.prog_bar_counter import ProgBarCounter
+from garage.sampler.base import BaseSampler
+
+
+class RaySampler(BaseSampler):
+    """Collects Policy Rollouts in a data parallel fashion.
+
+    Args:
+        - algo: A garage algo object
+        - env: A gym/akro env object
+        - should_render(bool): should the sampler render the trajectories
+        - sampler_worker_cls: If none, uses the default SamplerWorker
+            class
+
+    """
+
+    def __init__(self,
+                 algo,
+                 env,
+                 seed,
+                 should_render=False,
+                 num_processors=None,
+                 sampler_worker_cls=None):
+        self._sampler_worker = ray.remote(SamplerWorker if sampler_worker_cls
+                                          is None else sampler_worker_cls)
+        self._env = env
+        self._algo = algo
+        self._seed = seed
+        deterministic.set_seed(self._seed)
+        self._max_path_length = self._algo.max_path_length
+        self._should_render = should_render
+        if not ray.is_initialized():
+            ray.init(ignore_reinit_error=True)
+        self._num_workers = num_processors if num_processors \
+            else psutil.cpu_count(logical=False)
+        self._all_workers = defaultdict(None)
+
+    def start_worker(self):
+        """Initialize a new ray worker."""
+        env_pkl_id = ray.put(pickle.dumps(self._env))
```

Comment:
doesn't `ray.put()` pickle the argument for you?

Reply:
I believe it does, I can go ahead and change this, and make sure that it doesn’t effect speed for any reasons.

Reply:
So it does pickle, but it tries to use cloud pickle by default when we don’t manually pickle the policy, and because the policy can’t be cloudpickled, that causes the sampler to fail.